[3단계] 블랙잭(베팅)

RoadtoS7 opened the pull request:
안녕하세요 리뷰어님
수익 계산이 생각보다 많이 쉽지 않네요.. ProfitCalculator의 로직이 생각보다 많이 복잡한 것 같아요... 😭 
항상 꼼꼼하게 피드백 주셔서 감사하빈다. 이번 단계도 잘 부탁드립니다!~

RoadtoS7 committed:
refactor: TrumpCard에서 allCards메서드의 파라미터 이름 변경

- 수정 이유
변경된 파라미터 타입의 이름에 따라서 파라미터의 이름도 바꾸어주었다.

RoadtoS7 committed:
refactor: CardScoreRule.idealScore(sumWhereAceOne:sumeWhereAceEleven)메서드 로직 리팩터링

ace를 11점으로 계산한 값이 21점을 넘지 않는다면 최적의 점수가 되기 때문에 이를 반환한다.
만일 21점을 넘는다면, ace를 1점으로 계산한 값이 최적의 점수가 될 수 밖에 없기 때문에 이를 반환하도록 수정했다.

RoadtoS7 committed:
doc: README에 2단계 요구사항 분석 추가

RoadtoS7 committed:
feat: Dealer(딜러)를 Player를 상속받아서 구현

- Dealer도 게임 참가자이기에 Player와 동일한 역할을 함으로 Player를 상속받아 구현했다
- Dealer의 이름이 '딜러'로 지정되어있으므로 이름없이 CardDeck만 입력받아서 객체를 생성할 수 있도록 생성자를 추가로 구현했다

RoadtoS7 committed:
feat: BlackjackAnswerReader가 카드 점수가 16이하인 딜러의 응답을 받을 때 유닛테스트 구현

RoadtoS7 committed:
feat: BlackjackAnswerReader가 카드 점수가 16이하인 딜러의 응답으로 yes를 반환하도록 구현

RoadtoS7 committed:
feat: BlackjackAnswerReader 리팩토링

RoadtoS7 committed:
feat: WinLoseResult에 대한 유닛 테스트 구현

RoadtoS7 committed:
feat: WinLoseResult의 failable intializer 구현

WinLoseResult는 winCount, loseCount로 양수만 가질 수 있다. 만일 intializer로 음수가 들어온다면 nil을 반환한다

RoadtoS7 committed:
refactor: WinLoseResult의 failable intializer 리팩터링

RoadtoS7 committed:
refactor: WinLoseResult가 어떤 Player의 승패결과인지 알 수 있도록 Player을 프로퍼티로 갖도록 수정

RoadtoS7 committed:
refactor: Dealer의 answer() 메서드 제거

- 제거 이유: 사용하지 않음

RoadtoS7 committed:
feat: 점수를 가지고 player의 승패 결과를 계산하는 BlackjackJudge의 유닛테스트 작성

RoadtoS7 committed:
feat: 점수를 가지고 player의 승패 결과를 계산하는 BlackjackJudge 구현

RoadtoS7 committed:
feat: BlackjackJudge.winLoseResult() 리팩토링

RoadtoS7 committed:
feat: main로직에서 dealer도 블랙젝 게임에 참여하도록 수정함

RoadtoS7 committed:
feat: 딜러가 카드를 수령했음을 알리는 OutputView.printDealerReceiveCard 메서드 구현

RoadtoS7 committed:
refactor: BlackjackStateDelegate.afterRecevieCard에서 딜러가 카드를 수령했을 때 메시지 출력하도록 수정

RoadtoS7 committed:
refactor: PlayerNameParser에서 로그 메시지 출력 구문 삭제

RoadtoS7 committed:
feat: WinLoseResult에서 특정 유저의 승패 기록을 꺼내는 subscript에 대한 유닛테스트 구현

RoadtoS7 committed:
feat: 게임 참가자들의 승패결과를 담는 WinLoseResults 구현

RoadtoS7 committed:
feat: 특정 참가자의 점수를 기준으로 참가자들의 승패 결과를 계산하는 BlackjackJudge.winLoseResults() 유닛 테스트 구현

RoadtoS7 committed:
feat: 특정 참가자의 점수를 기준으로 참가자들의 승패 결과를 계산하는 BlackjackJudge.winLoseResults() 구현

RoadtoS7 committed:
feat: BlackjackJudge.winLoseResults(with:comparingWith:) 리팩터링

RoadtoS7 committed:
refaactor: BlackjackJudge.winLoseResults의 with 파라미터를 of 로 네이밍 변경

RoadtoS7 committed:
fix: WinLoseResults의 subscript 구현이 이전 커밋에 포함되지 않아서 발생하는 이슈 해결

RoadtoS7 committed:
feat: OutputView에서 최종 승패를 출력하는 메서드 구현

RoadtoS7 committed:
feat: WinLoseResult를 출력물로 만들어주는 WinLoseResultFormatter 구현

RoadtoS7 committed:
refactor: main에 승패결과를 계산해서 출력하는 로직 추가

RoadtoS7 committed:
refactor: 출력할 때 줄 간격 수정

RoadtoS7 committed:
feat: BlackjackGameJudge에서 딜러의 점수가 21점을 넘어가면 나머지 참가자들이 무조건 우승하는 유닛테스트 구현

RoadtoS7 committed:
refactor: BlackjackGameJudge.winLoseResult()에서 승패수를 구할 때 딜러의 점수가 21점을 넘으면 무조건 참가자들이 우승하는 것으로 수정

RoadtoS7 committed:
refactor: BlackjackGameJudge.winLoseResult() 리팩토링

RoadtoS7 committed:
refactor: 모든 forammter를 Util 디렉토리로 이동

RoadtoS7 committed:
refactor: WinLoseFormatter를 WinLoseResultFormatter로 네이밍 변경

RoadtoS7 committed:
feat: BlackjackScore의 유닛테스트 구현

RoadtoS7 committed:
feat: BlackjackScores.playerScores의 유닛 테스트 구현

- playerScores는 딜러를 제외한 참가자들의 점수만 리턴한다.

RoadtoS7 committed:
feat: BlackjackScores.playerScores 로직 구현

RoadtoS7 committed:
feat: BlackjackScorese.dealerScore 구현

RoadtoS7 committed:
feat: BlackjackGame.start()의 결과로 BlackjackScores를 반환받도록 구현

RoadtoS7 committed:
refactor: WinLoseResults에서 subsciprt를 제거하고, player점수와 딜러의 점수를 프로퍼티로 구분해서 갖도록 수정

RoadtoS7 committed:
refactor: BlackjackGameJudge의 메서드 파라미터 명, 변수 명을 변경하여 가독성 개선

RoadtoS7 committed:
style: WinLoseResult에서 네이밍과 괄호 사이에 띄어쓰기 함

RoadtoS7 committed:
refactor: WinLoseResult에서 zero 상수 대신 Int.zero를 사용하도록 수정

RoadtoS7 committed:
refactor: winLoseResults()메서드에 대한 유닛테스트 이름 수정

- 수정 이유

승패 결과를 반환하는 것인데, 실패 결과를 반환하는 것이라고 오타가 났었다.

RoadtoS7 committed:
refactor: BlackjackScoreTest에서 dealerScore에 대한 유닛테스트 이름에 존재하는 오타 수정

RoadtoS7 committed:
style: main로직의 인덴테이션 맞춤

RoadtoS7 committed:
style: Card, AnswerParser 파일의 인덴테이션 맞춤

RoadtoS7 committed:
fix: NonDistributeCardDistributor에서 중복된 카드가 뽑히는 이슈 해결

- 이슈 원인
distribute()메서드에서 여러장의 카드를 출력할 때 (count 파라미터 값이 2이상인 경우), 뽑은 카드를 카드 집합에서 제거하지 않기 때문에
중복된 카드가 뽑힐 수 있다.

- 이슈 해결 방법
count 만큼 반복문을 돌며 카드를 뽑을 때, 카드를 뽑자마자 뽑은 카드를 카드 집합에서 제거하도록 했다.

RoadtoS7 committed:
refactor: BlackjackGameTest에서 오타 수정

RoadtoS7 committed:
refactor: 'test_BlackjackScores_dealerScore_딜러의_점수를_반환한다'에서 playerScore와 일치하지 않는지 테스트하도록 수정

RoadtoS7 committed:
refactor: Player에서 Hashable 제거, Equatable 채택하도록 수정

RoadtoS7 committed:
refactor: WinLoseResult에서 Int.zero -> .zero로 줄임

RoadtoS7 committed:
fix: 오타로 인한 컴파일 에러 수정

RoadtoS7 committed:
feat: InputView에 플레이어의 배팅금액 입력받는 메서드 구현

RoadtoS7 committed:
feat: 음수를 받으면 nil을 반환하는지 테스트하는 MoneyTest 구현

RoadtoS7 committed:
feat: Money 구현

RoadtoS7 committed:
feat: 나누기는_두_moneyValue값을_나눈_몫에해당하는_정수를_반환한다 moneyTest 구현

RoadtoS7 committed:
feat: Money 나눗셈 구현

RoadtoS7 committed:
feat: Money 곱하기 연산 테스트 구현

RoadtoS7 committed:
feat: Money 곱하기 연산 구현

RoadtoS7 committed:
feat: 문자열 입력을 Money 인스턴스로 파싱하는 Money Parser 테스트 구현

RoadtoS7 committed:
feat: MoneyParser 구현

RoadtoS7 committed:
feat: MoneyParserTest에 nil이_입력되면_에러를_발생시키는지 체크하는 유닛 테스트 구현

RoadtoS7 committed:
feat: MoneyParserTest에 음수가 입력되면 에러를 발생시키는지 체크하는 유닛 테스트 구현

RoadtoS7 committed:
feat: MoneyParserTest에 숫자가 아닌 값이 입력되면 에러를 발생시키는지 체크하는 유닛테스트 구현

RoadtoS7 committed:
feat: ProfitCalculator 테스트 구현

RoadtoS7 committed:
feat: 참가자의 수익을 나타내는 Profit 구현

RoadtoS7 committed:
refactor: Money intializer에서 파라미터 이름 제거

RoadtoS7 committed:
feat: Money.zero 인스턴스 프로퍼티 구현

RoadtoS7 committed:
feat: Profit의 일급 컬렉션인 Profits 구현

RoadtoS7 committed:
refactor: Player가 bettingMoney를 갖도록 수정

RoadtoS7 committed:
refactor: ProfitCalculatorTest에서 참가자는_베팅금액을_21_초과할경우_베팅금액을_모두_잃는다 유닛테스트 로직 수정

RoadtoS7 committed:
feat: Money의 더하기 연산에 대한 유닛테스트 구현

RoadtoS7 committed:
feat: Money의 더하기 연산 구현

RoadtoS7 committed:
feat: Money의 더하기 연산 구현

RoadtoS7 committed:
feat: Money의 더하기 연산 구현

RoadtoS7 committed:
fix: BlackjackGameJudge가 player의 점수가 21을 초과해도 '승'으로 판정하는 이슈 수정

RoadtoS7 committed:
refactor: Money의 곱하기 연산 결과로 Int 반환하도록 수정

RoadtoS7 committed:
refactor: Player가 bettingMoney를 갖도록 수정

RoadtoS7 committed:
refactor: ProfitCalculator가 유닛테스트 통과하도록 구현

RoadtoS7 committed:
refactor: Money를 Equatable하게 수정

RoadtoS7 committed:
refactor: Profit 타입을 프로토콜로 빼고, PlusProfit, MinusProfit, ZeroProfit 구조체를 각각 구현

RoadtoS7 committed:
refactor: Money 나눗셈 유닛 테스트 제거

- 제거 이유: 사용안함

RoadtoS7 committed:
refactor: ProfitCalculatorTest의 기존 유닛 테스트 전부 제거, 딜러만_블랙잭일때_참가자들은_베팅금액을_모두_잃는다 유닛테스트 구현

RoadtoS7 committed:
refactor: ProfitCalculator의 기존 메서드 제거 및 딜러만_블랙잭일때_참가자들은_베팅금액을_모두_잃는다 유닛테스트 통과하도록 구현

RoadtoS7 committed:
feat: isBlackjack() 유닛테스트 구현

RoadtoS7 committed:
feat: isBlackjack() 구현

RoadtoS7 committed:
feat: Player.isBlackjack 테스트 구현

RoadtoS7 committed:
feat: Player.isBlackjack 구현

RoadtoS7 committed:
refactor: ProfitCalcultor 리팩터링

RoadtoS7 committed:
feat: ProfitCalculator 딜러와_참가자_둘다_블랙잭일때_참가자는_베팅금액을_돌려받아서_수익이_0이다 유닛 테스트 구현

RoadtoS7 committed:
feat: ProfitCalculator 딜러가_블랙잭이_아닐때_블랙잭_player는_배팅금액의_일점오배를_수익으로_얻는다 유닛테스트 구현

RoadtoS7 committed:
feat: Money를 상수와 곱할 수 있는 메서드 추가

RoadtoS7 committed:
feat: Profit간에 더하기, 빼기 가능하도록 구현

RoadtoS7 committed:
feat: ProfitCalculator 딜러가_블랙잭이_아닐때_블랙잭_player는_배팅금액의_일점오배를_수익으로_얻는다 기능 구현

RoadtoS7 committed:
refactor: ProfitCalcultor 리팩터링

RoadtoS7 committed:
feat: Profits 인스턴스를 출력물 형태로 포맷팅하는 ProfitsFormatter 구현

RoadtoS7 committed:
feat: OutputView에 최종 수익을 출력하는 메서드 구현

RoadtoS7 committed:
fix: ProfitsFormatter가 음수인 경우에는 수익에 -를 붙이도록 수정

RoadtoS7 committed:
refactor: InputView.readBettingMoney의 파라미터를 PlayerName으로 수정

RoadtoS7 committed:
refactor: main에서 배팅금액 받아서 최종 수익 출력하도록 수정

RoadtoS7 committed:
Resolve Conflict with sprint/2022-03

zdodev commented on Blackjack/Money.swift:
```diff
@@ -0,0 +1,32 @@
+import Foundation
+
+struct Money: Equatable {
+    static func * (lhs: Money, rhs: NSNumber) -> Money? {
+        let value = Double(lhs.value) * Double(truncating: rhs)
+        return Money(Int(value))
+    }
```
블랙잭 베팅 결과 금액인 1.5배 연산을 위해서 구현하신 부분일까요?
곱하기 시 금액을 1.5가 아닌 실수로 다른 숫자를 입력할 경우 로직과 다른 결과를 나타낼 수 있지 않을까 싶습니다!
실수를 방지하기 위해 1.5를 입력하지 않고 1.5배를 한 결과를 반환하도록 구현하는 것은 어떨까요? 의견이 궁금합니다! 지금도 좋습니다.👍

RoadtoS7 replied:
게임 참가자가 블랙잭일 때 우승 금액(베팅 금액의 1.5배)를 구하는 것이 목적이기 때문에, 리뷰어님이 피드백 주신대로 1.5배한 결과를 반환하는 것이 더 적합한 것 같습니다!
[3ec32fb](https://github.com/next-step/swift-blackjack/pull/14/commits/3ec32fbdfe8339b6ba31ee3e0a06f76a32058ec7) 커밋에서 반영했습니다! 피드백 감사합니다!!~

zdodev commented on Blackjack/Money.swift:
```diff
@@ -0,0 +1,32 @@
+        return Money(Int(value))
+    }
+    
+    static func * (lhs: Money, rhs: Money) -> Int {
```
곱하기 연산자를 구현하신 이유가 있을까요?

RoadtoS7 replied:
블랙잭 베팅금액의 1.5배를 구할 때 만들어놓고 사용하지 않았습니다..!  [4a2e94c](https://github.com/next-step/swift-blackjack/pull/14/commits/4a2e94c5e9a9a4e37f0548abcd900558e1764159) 커밋에서 제거했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/Money.swift:
```diff
@@ -0,0 +1,32 @@
+    static let zero = Money(.zero)!
+    let value: Int
+    
+    init?(_ value: Int) {
+        if value < .zero { return nil }
+        self.value = value
+    }
```
`Money` 타입을 항상 양수로 생성하도록 하신 이유가 궁금합니다!
`Profit` 타입을 `Plus`, `Minus`, `Zero` 로 나눈 것과 연관이 있는지도 궁금합니다!

RoadtoS7 replied:
음수의 `돈` 이라는 것은 현실에서 존재하지 않고, 
'수익/이득이 마이너스다/플러스다'라는 개념만 현실에 존재한다고 생각했습니다!
그래서 돈에 해당하는 Money는 항상 양수만 갖도록 하고,  Profit을 Plus, Minus, Zero로 나누어서 가질 수 있도록 구현했습니다!

zdodev replied:
`돈`이라는 것은 양수의 가치만 가지고, 수익에 대해서는 플러스, 마이너스, 0을 표현해 보신 거네요. 
이러한 표현법에 대한 의도와 생각이 궁금했습니다! 좋습니다!👍

zdodev replied:
`돈`이라는 것은 양수의 가치만 가지고, 수익에 대해서는 플러스, 마이너스, 0을 표현해 보신 거네요. 
이러한 표현법에 대한 의도와 생각이 궁금했습니다! 좋습니다!👍

zdodev commented on Blackjack/ProfitCalculator.swift:
```diff
@@ -0,0 +1,99 @@
+        return profitWith(nonBlackjackDealer: dealer, using: playerResults)
+    }
+    
+    static func profitWith(blackjackDealer: Player, players: [Player]) -> Profits {
```
외부에서 사용하지 않는 메서드로 보입니다!

RoadtoS7 replied:
[fac7b7e](https://github.com/next-step/swift-blackjack/pull/14/commits/fac7b7e725a139c4bcdcf90ba4544be3cbf42def) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/ProfitCalculator.swift:
```diff
@@ -0,0 +1,99 @@
+        return Profits(value: [dealerProfit] + profitsOfBlackjack + profitsOfNonBlackjack)
+    }
+    
+    static func profitWith(nonBlackjackDealer dealer: Player, using playerResults: [WinLoseResult]) -> Profits {
```
외부에서 사용하지 않는 메서드로 보입니다!

RoadtoS7 replied:
[fac7b7e](https://github.com/next-step/swift-blackjack/pull/14/commits/fac7b7e725a139c4bcdcf90ba4544be3cbf42def) 커밋에서 반영했습니다! 피드백 감사합니다!

zdodev commented on Blackjack/ProfitCalculator.swift:
```diff
@@ -0,0 +1,99 @@
+        return Profits(value: [dealerProfit] + winPlayersProfits + losePlayersProfits + blackjackPlayersProfit)
+    }
+    
+    static func sum(player: Player, _ profits: [Profit]) -> Profit {
```
외부에서 사용하지 않는 메서드로 보입니다!

RoadtoS7 replied:
[fac7b7e](https://github.com/next-step/swift-blackjack/pull/14/commits/fac7b7e725a139c4bcdcf90ba4544be3cbf42def) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev reviewed (changes requested):
안녕하세요! @RoadtoS7 

블랙잭 3단계도 기능 요구 사항과 프로그래밍 요구 사항을 잘 지켜지도록 잘 구현해주셨습니다!👍

약간의 궁금증을 남겨보았습니다.
같이 의견 나눠보면 좋을 것 같습니다!

RoadtoS7 committed:
refactor: Money끼리 곱하여 Money를 반환하는 메서드 제거

RoadtoS7 committed:
refactor: ProfitCaculator에서 외부에서 사용하지 않는 메서드는 private이 되도록 수정함

RoadtoS7 committed:
refactor: Money에 숫자를 곱한 결과를 Money로 얻을 수 있는 메서드를 제거하고, blackjackWinning()메서드를 사용하도록 수정함

- 수정 이유
결국 상수와 곱하기 연산을 구현한 이유가 참가자가 blackjack일 때 자신이 베팅한 금액의 1.5배를 얻을 수 있도록 하기 위함이다.
따라서 블랙잭일 때 얻을 수 있는 금액을 구하는 것이 목적이므로, 이 기능을 수행하는 것을 메서드로 구현했다

RoadtoS7 committed:
fix: 오타로 인한 에러 수정

zdodev commented on Blackjack/Money.swift:
```diff
@@ -0,0 +1,29 @@
+        self.value = value
+    }
+    
+    func blackjackWinningMoney() -> Money {
+        let winningMoneyValue = Int(Double(value) * 1.5)
+        return Money(winningMoneyValue)!
+    }
```
좋습니다!👍

zdodev reviewed (approved):
안녕하세요! @RoadtoS7 

블랙잭 프로젝트 베팅까지 잘 구현해주셨습니다!
구현 요구 사항이 쉽지 않았지만, 요구 사항을 잘 지켜주시고 
테스트 코드까지 적절하게 잘 작성해주신 점이 좋았습니다!👍👍
짧은 기간동안 꾸준히 PR을 보내주시고 잘 진행해 주셨습니다!!👏👏

그동안 블랙잭 프로젝트를 진행하시느라 고생 많으셨습니다.🎉🎉
